destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala

```diff
@@ -0,0 +1,221 @@
+    def drop: ConnectionIO[Unit]
+    def create: ConnectionIO[Unit]
+    def persist(writeMode: WriteMode): ConnectionIO[Unit]
+    def append: ConnectionIO[Unit]
+  }
+
+  private object TempTable {
+    def apply(
+        log: LogHandler,
+        unsafeName: String,
+        unsafeSchema: Option[String],
+        tableFragment: Fragment,
+        columns: NonEmptyList[(HI, SQLServerType)],
+        idColumn: Option[Column[_]])
+        : TempTable = {
+      val schema = unsafeSchema.getOrElse("dbo")
+      val tempName = s"precog_temp_$unsafeName"
+      val hyName = SQLServerHygiene.hygienicIdent(Ident(tempName))
+      val hySchema = SQLServerHygiene.hygienicIdent(Ident(schema))
+      val tempFragment = Fragment.const0(hySchema.forSqlName) ++ fr0"." ++ Fragment.const0(hyName.forSqlName)
+      val uName = hyName.unsafeForSqlName
+      val uSchema = hySchema.unsafeForSqlName
+
+      new TempTable {
+        def ingest(chunk: Chunk[CharSequence]): ConnectionIO[Unit] =
+          insertChunk(log)(tempFragment, columns, chunk)
+
+        def drop: ConnectionIO[Unit] =
+          ifExists(log)(uName, uSchema.some).option flatMap { results =>
+            if (results.exists(_ === 1)) {
+              (fr"DROP TABLE" ++ tempFragment)
+                .updateWithLogHandler(log)
+                .run
+                .void
+            } else {
+              ().pure[ConnectionIO]
+            }
+          }
+
+        def create: ConnectionIO[Unit] =
+          ifExists(log)(uName, uSchema.some).option flatMap { results =>
+            if (results.exists(_ === 0)) {
+              val columnsObj = createColumnSpecs(columns)
+              (fr"CREATE TABLE" ++ tempFragment ++ fr0" " ++ columnsObj)
+                .updateWithLogHandler(log)
+                .run
+                .void
+            } else {
+              ().pure[ConnectionIO]
+            }
+          }
+
+        private def truncate: ConnectionIO[Unit] =
+          ifExists(log)(uName, uSchema.some).option flatMap { results =>
+            if (results.exists(_ === 1)) {
+              (fr"TRUNCATE TABLE" ++ tempFragment)
+                .updateWithLogHandler(log)
+                .run
+                .void
+            } else {
+              ().pure[ConnectionIO]
+            }
+          }
+
+        private def insertInto: ConnectionIO[Unit] =
+          (fr"INSERT INTO" ++
+            tableFragment ++ fr0" " ++
+            fr"SELECT * FROM" ++
+            tempFragment)
+            .updateWithLogHandler(log)
+            .run
+            .void
+
+        private def rename: ConnectionIO[Unit] =
+          (fr0"EXEC SP_RENAME '" ++
+            tempFragment ++ fr0"', '" ++
+            Fragment.const0(SQLServerHygiene.hygienicIdent(Ident(unsafeName)).unsafeForSqlName) ++ fr0"'")
+            .updateWithLogHandler(log)
+            .run
+            .void
+
+        private def filterTempIds(idColumn: Column[_]): ConnectionIO[Unit] = {
```

Comment:
Might make sense to index the temp table for this join, though we'd want to see some query plans/test it to be sure.

Reply:
It wouldn't harm I suppose. Target table has index and temp one should be small.